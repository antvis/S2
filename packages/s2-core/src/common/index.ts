export * from './constant';
export * from './debug';
export * from './i18n';
export * from './icons';
export * from './interface';
export * from './store';
