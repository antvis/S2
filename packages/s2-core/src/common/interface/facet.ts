import type { Hierarchy } from '../../facet/layout/hierarchy';
import type { Node } from '../../facet/layout/node';
import type { ViewMeta } from './basic';

export interface AdjustLeafNodesParams {
  leafNodes: Node[];
  hierarchy: Hierarchy;
}

export interface ScrollChangeParams {
  offset: number;
  updateThumbOffset: boolean;
}

export type GetCellMeta = (
  rowIndex: number,
  colIndex: number,
) => ViewMeta | null;

export interface LayoutResult {
  /**
   * 列头节点, 对应 ColCell (含可视范围外)
   */
  colNodes: Node[];
  /**
   * 列头叶子节点, 对应 ColCell (含可视范围外)
   */
  colLeafNodes: Node[];
  /**
   * 列头节点层级结构 (含可视范围外)
   */
  colsHierarchy: Hierarchy;
  /**
   * 行头节点, 对应 RowCell (含可视范围外)
   */
  rowNodes: Node[];
  /**
   * 行头节点层级结构 (含可视范围外)
   */
  rowsHierarchy: Hierarchy;
  /**
   * 行头叶子节点, 对应 RowCell (含可视范围外)
   */
  rowLeafNodes: Node[];
  /**
   * 序号节点, 对应 SeriesNumberCell (含可视范围外)
   */
  seriesNumberNodes?: Node[];
  /**
   * 角头节点, 对应 CornerCell (含可视范围外)
   */
  cornerNodes?: Node[];
}
