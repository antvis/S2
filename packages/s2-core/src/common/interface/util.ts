export type DeepRequired<T extends Record<string, any>> = {
  [K in keyof T]-?: NonNullable<T[K]> extends Record<string, any>
    ? DeepRequired<NonNullable<T[K]>>
    : NonNullable<T[K]>;
};
