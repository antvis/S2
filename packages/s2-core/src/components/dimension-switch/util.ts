import { DimensionType, DimensionItem } from './dimension';

export interface OperatedType {
  [type: string]: string[];
}

export const getDimensionsByPredicate = (
  data: DimensionType[],
  predicate: (value: DimensionItem) => boolean,
) => {
  return data.reduce((obj, t) => {
    obj[t.type] = t.items.filter(predicate).map((i) => i.id);
    return obj;
  }, {} as OperatedType);
};
