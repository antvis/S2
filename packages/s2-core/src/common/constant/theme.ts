import { Palette } from '../interface';
import { paletteColorful } from '@/theme/palette/colorful';
import { paletteDefault } from '@/theme/palette/default';
import { paletteGray } from '@/theme/palette/gray';

// Map of the theme
export const PALETTE_MAP: Record<string, Palette> = {
  default: paletteDefault,
  colorful: paletteColorful,
  gray: paletteGray,
};

export const FONT_FAMILY =
  'Roboto, PingFangSC, BlinkMacSystemFont, Microsoft YaHei, Arial, sans-serif';

export const MINI_BAR_CHART_HEIGHT = 12;
