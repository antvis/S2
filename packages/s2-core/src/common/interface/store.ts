import { BBox, Group } from '@antv/g-canvas';
import {
  InteractionStateInfo,
  S2CellType,
  SortParam,
  ViewMeta,
} from '../interface';
import { ValueRanges } from './condition';
import { S2DataConfig } from './s2DataConfig';
import { Node } from '@/facet/layout/node';
import { PartDrillDownFieldInLevel } from '@/common/interface';
import { GuiIcon } from '@/common/icons';

export interface Selected {
  type:
    | 'cell'
    | 'row'
    | 'brush'
    | 'col'
    | 'column'
    | 'row&col'
    | 'column-multiple';
  // [ 10, 5 ], [ [ 2, 5 ], [ 4, 8 ] ];
  indexes: [number | number[], number | number[]];
}

export interface ReachedBorderId {
  rowId?: string;
  colId?: string;
}

export interface HiddenColumnsInfo {
  // 当前显示的兄弟节点之前所隐藏的节点
  hideColumnNodes: Node[];
  // 当前隐藏列所对应展示展开按钮的兄弟节点
  displaySiblingNode: {
    prev: Node;
    next: Node;
  };
}

/**
 * All stored keys need type define here
 */
export interface StoreKey {
  // horizontal scroll bar scroll x offset
  scrollX: number;
  // vertical scroll bar scroll y offset
  scrollY: number;
  // row header scroll bar scroll x offset
  hRowScrollX: number;
  // column cell click sort params
  sortParam: SortParam;
  // 下钻节点id和对应生成的 path寻址路径
  drillDownIdPathMap: Map<string, number[][]>;
  // 当前下钻节点
  drillDownNode: Node;
  // 下钻数据的个数控制
  drillItemsNum: number;
  drillDownFieldInLevel: PartDrillDownFieldInLevel[];

  // interaction state
  interactionStateInfo: InteractionStateInfo;

  originalDataCfg: S2DataConfig;

  panelBBox: BBox;
  // resize area group
  activeResizeArea: Group;
  // interval condition
  valueRanges: ValueRanges;
  // 初次渲染时的列头叶子节点
  initColumnLeafNodes: Node[];
  /**
   * 隐藏列详情
   *  | a, b, [c,d 隐藏] [icon e ] , [f 隐藏], [icon g]   |
   */
  hiddenColumnsDetail: HiddenColumnsInfo[];
  // 上一次渲染的列头配置
  lastRenderedColumnFields: string[];
  // 是否手动调整过宽高
  resized: boolean;

  // hover 显示的 icon 缓存
  visibleActionIcons: GuiIcon[];

  // last click cell
  lastClickedCell: S2CellType<ViewMeta>;

  [key: string]: unknown;
}
