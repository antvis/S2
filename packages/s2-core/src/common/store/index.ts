import { StoreKey } from '@/common/interface';

/**
 * Store something in {@link SpreadSheet} temporary along with it's lifecycle
 * All the keys need be declare in {@see StoreKey} first
 */
export class Store {
  private store = new Map<keyof StoreKey, unknown>();

  public set<T extends keyof StoreKey>(key: T, value: StoreKey[T]) {
    this.store.set(key, value);
  }

  public get<T extends keyof StoreKey>(
    key: T,
    defaultValue?: StoreKey[T],
  ): StoreKey[T] {
    const value = this.store.get(key);
    return (value as StoreKey[T]) ?? defaultValue;
  }

  public clear() {
    this.store.clear();
  }

  public size() {
    return this.store.size;
  }
}
