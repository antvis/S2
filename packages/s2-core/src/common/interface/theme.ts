import { ShapeAttrs } from '@antv/g-canvas';
import { InteractionStateName } from '../constant';
import { CellTypes } from '@/common/constant/interaction';

// 文本内容的水平对齐方式, 默认 left
export type TextAlign = 'left' | 'center' | 'right';

// 绘制文本时的基线, 对应垂直方向对齐方式 默认 bottom
export type TextBaseline = 'top' | 'middle' | 'bottom';

export interface Palette {
  /* 基础色值 */
  basicColors: string[];
  /* 语义色值 */
  semanticColors: {
    red?: string;
    green?: string;
    /* 额外颜色字段 */
    [key: string]: string;
  };
}

export interface Padding {
  top?: number;
  right?: number;
  bottom?: number;
  left?: number;
}

export interface Background {
  opacity?: number;
  color?: string;
}
export interface InteractionStateTheme {
  /* 背景透明度 */
  backgroundOpacity?: number;
  /* 背景填充色 */
  backgroundColor?: string;
  /* 文本透明度 */
  textOpacity?: number;
  /* 边线颜色 */
  borderColor?: string;
  /* 边线宽度 */
  borderWidth?: number;
  /* 边线透明度 */
  borderOpacity?: number;
  /* 透明度 */
  opacity?: number;
}

export type InteractionState = {
  [K in InteractionStateName]?: InteractionStateTheme;
};

export type Margin = Padding;

export interface TextAlignCfg {
  textAlign?: TextAlign;
  textBaseline?: TextBaseline;
}

export interface TextTheme extends TextAlignCfg {
  /* 字体 */
  fontFamily?: string;
  /* 字体大小 */
  fontSize?: number;
  /* 字体粗细 */
  fontWeight?: number | 'normal' | 'bold' | 'bolder' | 'lighter';
  /* 字体颜色 */
  fill?: string;
  /* 链接文本颜色 */
  linkTextFill?: string;
  /* 字体透明度 */
  opacity?: number;
}

export interface CellTheme {
  /* 奇数行单元格背景色 */
  crossBackgroundColor?: string;
  /* 单元格背景色 */
  backgroundColor?: string;
  /* 单元格背景色透明度 */
  backgroundColorOpacity?: number;
  /* 单元格水平边线颜色 */
  horizontalBorderColor?: string;
  /* 单元格水平边线颜色透明度 */
  horizontalBorderColorOpacity?: number;
  /* 单元格垂直边线颜色 */
  verticalBorderColor?: string;
  /* 单元格垂直边线颜色透明度 */
  verticalBorderColorOpacity?: number;
  /* 单元格水平边线宽度 */
  horizontalBorderWidth?: number;
  /* 单元格垂直边线宽度 */
  verticalBorderWidth?: number;
  /* 单元格内边距 */
  padding: Padding;
  /* 交互态 */
  interactionState?: InteractionState;
  /* 单元格内条件格式-迷你条形图高度 */
  miniBarChartHeight?: number;
  /* 单元格内条件格式-迷你条形图默认填充颜色 */
  miniBarChartFillColor?: string;
}

export interface IconTheme {
  /* icon 填充色 */
  fill?: string;
  /* 下跌 icon 填充色 */
  downIconColor?: string;
  /* 上涨 icon 填充色 */
  upIconColor?: string;
  /* icon  大小 */
  size?: number;
  /* icon 外边距 */
  margin?: Margin;
}

export interface ResizeArea {
  /* 热区尺寸 */
  size?: number;
  /* 热区背景色 */
  background?: string;
  /* 参考线颜色 */
  guideLineColor?: string;
  /* 参考线间隔 */
  guideLineDash?: number[];
  /* 热区背景色透明度 */
  backgroundOpacity?: number;
  /* 交互态 */
  interactionState?: InteractionState;
}

export interface ScrollBarTheme {
  /* 滚动条轨道颜色 */
  trackColor?: string;
  /* 滚动条 hover 态颜色 */
  thumbHoverColor?: string;
  /* 滚动条颜色 */
  thumbColor?: string;
  /* 滚动条尺寸 */
  size?: number;
  /* 滚动条 hover 态尺寸 */
  hoverSize?: number;
  /** 指定如何绘制每一条线段末端，lineCap?: 'butt' | 'round' | 'square'; */
  lineCap?: ShapeAttrs['lineCap'];
}

export interface SplitLine {
  /* 水平分割线颜色 */
  horizontalBorderColor?: string;
  /* 水平分割线颜色透明度 */
  horizontalBorderColorOpacity?: number;
  /* 水平分割线宽度 */
  horizontalBorderWidth?: number;
  /* 垂直分割线颜色 */
  verticalBorderColor?: string;
  /* 垂直分割线颜色透明度 */
  verticalBorderColorOpacity?: number;
  /* 垂直分割线宽度 */
  verticalBorderWidth?: number;
  /* 分割线是否显示外阴影 */
  showShadow?: boolean;
  /* 阴影宽度 */
  shadowWidth?: number;
  /* 阴影线性渐变色 */
  shadowColors?: {
    /* 线性变化左侧颜色 */
    left: string;
    /* 线性变化右侧颜色 */
    right: string;
  };
}
export interface DefaultCellTheme extends GridAnalysisCellTheme {
  /* 粗体文本样式 */
  bolderText?: TextTheme;
  /* 文本样式 */
  text?: TextTheme;
  /* 单元格样式 */
  cell?: CellTheme;
  /* 图标样式 */
  icon?: IconTheme;
  /* 序号列宽 */
  seriesNumberWidth?: number;
}

export interface GridAnalysisCellTheme {
  // 次级文本，如副指标
  minorText?: TextTheme;
  // 衍生指标
  derivedMeasureText?: {
    mainUp: string;
    mainDown: string;
    minorUp: string;
    minorDown: string;
  };
}

export type CellThemes = {
  [K in CellTypes]?: DefaultCellTheme;
};

export interface S2Theme extends CellThemes {
  /* 列宽行高调整热区 */
  resizeArea?: ResizeArea;
  /* 滚动条样式 */
  scrollBar?: ScrollBarTheme;
  /* 分割线样式 */
  splitLine?: SplitLine;
  /* 刷选遮罩 */
  prepareSelectMask?: InteractionStateTheme;
  /* 画布背景底色 */
  background?: Background;
}

export type ThemeName = 'default' | 'colorful' | 'gray';

export interface ThemeCfg {
  /* 主题 */
  theme?: S2Theme;
  /* 色板 */
  palette?: Palette;
  /* 主题名 */
  name?: ThemeName;
}
