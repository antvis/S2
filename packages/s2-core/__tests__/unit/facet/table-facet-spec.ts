/**
 * table mode pivot test.
 */
import { Canvas } from '@antv/g-canvas';
import { assembleDataCfg, assembleOptions } from 'tests/util';
import { merge } from 'lodash';
import { data } from '../../data/mock-dataset.json';
import { SpreadSheet } from '@/sheet-type';
import { TableDataSet } from '@/data-set/table-data-set';
import { TableFacet } from '@/facet/table-facet';
import { Store } from '@/common/store';
import { getTheme } from '@/theme';
import { DEFAULT_STYLE } from '@/index';

jest.mock('src/sheet-type', () => {
  const container = new Canvas({
    width: 100,
    height: 100,
    container: document.body,
  });
  return {
    SpreadSheet: jest.fn().mockImplementation(() => {
      return {
        dataCfg: {
          ...assembleDataCfg(),
        },
        options: assembleOptions(),
        panelScrollGroup: {
          setClip: jest.fn(),
        },
        container,
        theme: getTheme({}),
        panelGroup: container.addGroup(),
        foregroundGroup: container.addGroup(),
        backgroundGroup: container.addGroup(),
        store: new Store(),
        on: jest.fn(),
        isTableMode: jest.fn().mockReturnValue(true),
        isPivotMode: jest.fn(),
        getTotalsConfig: jest.fn(),
        getLayoutWidthType: jest.fn().mockRejectedValue('adaptive'),
        emit: jest.fn(),
        isScrollContainsRowHeader: jest.fn(),
        getColumnLeafNodes: jest.fn().mockReturnValue([]),
        isHierarchyTreeType: jest.fn(),
      };
    }),
  };
});
jest.mock('src/data-set/table-data-set', () => {
  return {
    TableDataSet: jest.fn().mockImplementation(() => {
      return {
        ...assembleDataCfg(),
        originData: data,
        displayData: data,
        moreThanOneValue: jest.fn(),
        getFieldName: jest.fn(),
        getDimensionValues: jest.fn(),
        getDisplayDataSet: jest.fn(() => data),
        getCellData: () => 1,
      };
    }),
  };
});
const MockSpreadSheet = SpreadSheet as any as jest.Mock<SpreadSheet>;
const MockTableDataSet = TableDataSet as any as jest.Mock<TableDataSet>;

describe('Table Mode Facet Test', () => {
  const ss: SpreadSheet = new MockSpreadSheet();
  const dataSet: TableDataSet = new MockTableDataSet(ss);
  const facet: TableFacet = new TableFacet({
    spreadsheet: ss,
    dataSet: dataSet,
    ...assembleDataCfg().fields,
    ...assembleOptions(),
    ...DEFAULT_STYLE,
    columns: ['province', 'city', 'type', 'sub_type', 'price'],
  });

  describe('should get correct row hierarchy', () => {
    const { rowsHierarchy } = facet.layoutResult;
    test('row hierarchy structure', () => {
      expect(rowsHierarchy.height).toBe(0);
      expect(rowsHierarchy.width).toBe(0);
      expect(rowsHierarchy.getIndexNodes()).toHaveLength(0);
    });
  });
});

describe('Table Mode Facet Test With Adaptive Layout', () => {
  const ss: SpreadSheet = new MockSpreadSheet();
  const dataSet: TableDataSet = new MockTableDataSet(ss);
  const options = {
    spreadsheet: ss,
    dataSet: dataSet,
    ...assembleDataCfg().fields,
    ...assembleOptions({}),
    ...DEFAULT_STYLE,
    layoutWidthType: 'adaptive',
    columns: ['province', 'city', 'type', 'sub_type', 'price'],
  };

  describe('should get correct col layout', () => {
    const facet: TableFacet = new TableFacet({
      ...options,
      showSeriesNumber: false,
    });
    const { colCfg, cellCfg } = facet.cfg;

    test('col hierarchy coordinate with adaptive layout', () => {
      const { colLeafNodes } = facet.layoutResult;

      const colHeaderColSize = colLeafNodes.length;
      const canvasW = facet.getCanvasHW().width;
      const adaptiveWith = Math.max(
        cellCfg.width,
        canvasW / Math.max(1, colHeaderColSize),
      );

      colLeafNodes.forEach((node, index) => {
        expect(node.y).toBe(0);
        expect(node.x).toBe(index * adaptiveWith);
        expect(Math.round(node.width)).toBe(adaptiveWith);
        expect(node.height).toBe(colCfg.height);
      });
    });
  });

  describe('should get correct col layout with seriesNumber', () => {
    const s2: SpreadSheet = new MockSpreadSheet();
    const s2DataSet: TableDataSet = new MockTableDataSet(s2);
    const facet = new TableFacet({
      ...options,
      spreadsheet: s2,
      dataSet: s2DataSet,
      showSeriesNumber: true,
    });
    const { colCfg, cellCfg } = facet.cfg;

    test('col hierarchy coordinate with adaptive layout with seriesNumber', () => {
      const { colLeafNodes } = facet.layoutResult;

      const seriesNumberWidth = facet.getSeriesNumberWidth();
      const colHeaderColSize = colLeafNodes.length - 1;
      const canvasW = facet.getCanvasHW().width - seriesNumberWidth;
      const adaptiveWith = Math.max(
        cellCfg.width,
        canvasW / Math.max(1, colHeaderColSize),
      );

      const seriesNumberNode = colLeafNodes[0];
      expect(seriesNumberNode.y).toBe(0);
      expect(seriesNumberNode.x).toBe(0);
      expect(seriesNumberNode.width).toBe(seriesNumberWidth);
      expect(seriesNumberNode.height).toBe(colCfg.height);

      colLeafNodes.slice(1).forEach((node, index) => {
        expect(node.y).toBe(0);
        expect(node.x).toBe(index * adaptiveWith + seriesNumberWidth);
        expect(node.width).toBe(adaptiveWith);
        expect(node.height).toBe(colCfg.height);
      });
    });
  });
});

describe('Table Mode Facet Test With Compact Layout', () => {
  describe('should get correct col layout', () => {
    const ss: SpreadSheet = new MockSpreadSheet();
    const dataSet: TableDataSet = new MockTableDataSet(ss);
    ss.getLayoutWidthType = () => {
      return 'compact';
    };
    const facet: TableFacet = new TableFacet({
      spreadsheet: ss,
      dataSet: dataSet,
      ...assembleDataCfg().fields,
      ...assembleOptions(),
      ...DEFAULT_STYLE,
      columns: ['province', 'city', 'type', 'sub_type', 'price'],
    });
    const { colCfg } = facet.cfg;

    test('col hierarchy coordinate with compact layout', () => {
      const { colLeafNodes } = facet.layoutResult;

      const COMPACT_WIDTH = [52, 52, 64, 40, 72];

      let lastX = 0;
      colLeafNodes.forEach((node, index) => {
        expect(node.y).toBe(0);
        expect(node.x).toBe(lastX);
        expect(Math.floor(node.width)).toEqual(COMPACT_WIDTH[index]);
        expect(node.height).toBe(colCfg.height);
        lastX += COMPACT_WIDTH[index];
      });
    });
  });

  describe('should get correct col layout with seriesNumber', () => {
    const ss: SpreadSheet = new MockSpreadSheet();
    const dataSet: TableDataSet = new MockTableDataSet(ss);
    ss.getLayoutWidthType = () => {
      return 'compact';
    };
    const facet: TableFacet = new TableFacet({
      spreadsheet: ss,
      dataSet: dataSet,
      ...assembleDataCfg().fields,
      ...assembleOptions(),
      ...DEFAULT_STYLE,
      columns: ['province', 'city', 'type', 'sub_type', 'price'],
      showSeriesNumber: true,
    });
    const { colCfg } = facet.cfg;

    test('col hierarchy coordinate with compact layout with seriesNumber', () => {
      const { colLeafNodes } = facet.layoutResult;

      const COMPACT_WIDTH = [80, 52, 52, 64, 40, 72];

      let lastX = 0;
      colLeafNodes.forEach((node, index) => {
        expect(node.y).toBe(0);
        expect(node.x).toBe(lastX);
        expect(Math.floor(node.width)).toBe(COMPACT_WIDTH[index]);
        expect(node.height).toBe(colCfg.height);
        lastX += COMPACT_WIDTH[index];
      });
    });
  });
});

describe('Table Mode Facet With Frozen Test', () => {
  const ss: SpreadSheet = new MockSpreadSheet();
  const dataSet: TableDataSet = new MockTableDataSet(ss);
  const facet: TableFacet = new TableFacet({
    spreadsheet: ss,
    dataSet: dataSet,
    ...assembleDataCfg().fields,
    ...assembleOptions({
      frozenColCount: 2,
      frozenRowCount: 2,
      frozenTrailingColCount: 2,
      frozenTrailingRowCount: 2,
    }),
    ...DEFAULT_STYLE,
    columns: ['province', 'city', 'type', 'sub_type', 'price'],
  });

  test('should get correct frozenInfo', () => {
    facet.calculateFrozenGroupInfo();
    expect(facet.frozenGroupInfo).toStrictEqual({
      col: {
        range: [0, 1],
        width: 240,
      },
      row: {
        height: 60,
        range: [0, 2],
      },
      trailingCol: {
        range: [3, 4],
        width: 240,
      },
      trailingRow: {
        height: 60,
        range: [29, 31],
      },
    });
  });

  test('should get correct xy indexes with frozen', () => {
    expect(facet.calculateXYIndexes(0, 0)).toStrictEqual({
      center: [2, 2, 2, 16],
      frozenCol: [0, 1, 2, 16],
      frozenRow: [2, 2, 0, 1],
      frozenTrailingCol: [3, 4, 2, 16],
      frozenTrailingRow: [2, 2, 30, 31],
    });

    expect(facet.calculateXYIndexes(100, 200)).toStrictEqual({
      center: [2, 2, 8, 23],
      frozenCol: [0, 1, 8, 23],
      frozenRow: [2, 2, 0, 1],
      frozenTrailingCol: [3, 4, 8, 23],
      frozenTrailingRow: [2, 2, 30, 31],
    });
  });

  test('should get correct col layout with frozen col', () => {
    const { width } = facet.spreadsheet.options;
    const { frozenTrailingColCount } = facet.cfg;
    const { colLeafNodes } = facet.layoutResult;
    let prevWidth = 0;
    colLeafNodes
      .slice(-frozenTrailingColCount)
      .reverse()
      .forEach((node) => {
        prevWidth += node.width;
        expect(node.x).toBe(width - prevWidth);
      });
  });

  test('should get correct cell layout with frozenTrailingCol', () => {
    const { width } = facet.spreadsheet.options;
    const { frozenTrailingColCount } = facet.cfg;
    const { colLeafNodes, getCellMeta } = facet.layoutResult;

    let prevWidth = 0;
    colLeafNodes
      .slice(-frozenTrailingColCount)
      .reverse()
      .forEach((node, index) => {
        prevWidth += node.width;
        expect(getCellMeta(1, colLeafNodes.length - 1 - index).x).toBe(
          width - prevWidth,
        );
      });
  });

  test('should get correct cell layout with frozenTrailingRow', () => {
    const { frozenTrailingRowCount, cellCfg } = facet.cfg;
    const { getCellMeta } = facet.layoutResult;
    const displayData = dataSet.getDisplayDataSet();
    const panelBBox = facet.panelBBox;
    let prevHeight = 0;
    displayData
      .slice(-frozenTrailingRowCount)
      .reverse()
      .forEach((_, idx) => {
        prevHeight += cellCfg.height;
        expect(getCellMeta(displayData.length - 1 - idx, 1).y).toBe(
          panelBBox.maxY - prevHeight,
        );
      });
  });

  test('should get correct viewCellHeights result', () => {
    const viewCellHeights = facet.getViewCellHeights();
    expect(viewCellHeights.getIndexRange(0, 715)).toStrictEqual({
      start: 0,
      end: 23,
    });
    expect(viewCellHeights.getIndexRange(1110, 1500)).toStrictEqual({
      start: 37,
      end: 49,
    });
    expect(viewCellHeights.getIndexRange(0, 0)).toStrictEqual({
      start: 0,
      end: 0,
    });

    expect(viewCellHeights.getTotalHeight()).toBe(960);
    expect(viewCellHeights.getTotalLength()).toBe(32);
    expect(viewCellHeights.getCellOffsetY(0)).toBe(0);
    expect(viewCellHeights.getCellOffsetY(7)).toBe(210);
  });
});

describe('Table Mode Facet Test With Custom Row Height', () => {
  const ss: SpreadSheet = new MockSpreadSheet();
  const dataSet: TableDataSet = new MockTableDataSet(ss);
  ss.options = merge({}, assembleOptions(), {
    style: {
      rowCfg: {
        heightByField: {
          '2': 300,
          '3': 200,
          '10': 60,
          '15': 20,
        },
      },
    },
  });
  const facet: TableFacet = new TableFacet({
    spreadsheet: ss,
    dataSet: dataSet,
    ...assembleDataCfg().fields,
    ...merge({}, assembleOptions()),
    ...DEFAULT_STYLE,
    columns: ['province', 'city', 'type', 'sub_type', 'price'],
  });

  test('should get correct rowOffsets when custom row height is set', () => {
    const rowOffsets = facet.rowOffsets;
    expect(rowOffsets).toStrictEqual([
      0, 30, 60, 360, 560, 590, 620, 650, 680, 710, 740, 800, 830, 860, 890,
      920, 940, 970, 1000, 1030, 1060, 1090, 1120, 1150, 1180, 1210, 1240, 1270,
      1300, 1330, 1360, 1390, 1420,
    ]);
  });

  test('should get correct viewCellHeights result when custom row height is set', () => {
    const viewCellHeights = facet.getViewCellHeights();

    expect(viewCellHeights.getIndexRange(0, 715)).toStrictEqual({
      start: 0,
      end: 9,
    });
    expect(viewCellHeights.getIndexRange(1110, 1500)).toStrictEqual({
      start: 21,
      end: 31,
    });

    expect(viewCellHeights.getTotalHeight()).toBe(1420);
    expect(viewCellHeights.getTotalLength()).toBe(32);
    expect(viewCellHeights.getCellOffsetY(0)).toBe(0);
    expect(viewCellHeights.getCellOffsetY(7)).toBe(650);
  });
});

describe('Table Mode Facet Test With Zero Height', () => {
  const ss: SpreadSheet = new MockSpreadSheet();
  const dataSet: TableDataSet = new MockTableDataSet(ss);
  ss.options = merge({}, assembleOptions(), {
    width: 0,
    height: 0,
  });
  const facet: TableFacet = new TableFacet({
    spreadsheet: ss,
    dataSet: dataSet,
    ...assembleDataCfg().fields,
    ...merge({}, assembleOptions()),
    ...DEFAULT_STYLE,
    columns: ['province', 'city', 'type', 'sub_type', 'price'],
  });

  test('should get correct panelBBox', () => {
    const panelBBox = facet.panelBBox;
    expect(panelBBox.width).toBe(0);
    expect(panelBBox.height).toBe(0);
  });

  test('should get correct initial scroll position', () => {
    const { scrollX, scrollY } = facet.getScrollOffset();
    expect(scrollX).toBe(0);
    expect(scrollY).toBe(0);
  });
});

describe('Table Mode Facet With Frozen layoutCoordinate Test', () => {
  const ss: SpreadSheet = new MockSpreadSheet();
  const dataSet: TableDataSet = new MockTableDataSet(ss);
  const facet: TableFacet = new TableFacet({
    spreadsheet: ss,
    dataSet: dataSet,
    ...assembleDataCfg().fields,
    ...assembleOptions({
      frozenColCount: 2,
      frozenRowCount: 2,
      frozenTrailingColCount: 2,
      frozenTrailingRowCount: 2,
    }),
    ...DEFAULT_STYLE,
    columns: ['province', 'city', 'type', 'sub_type', 'price'],
    layoutCoordinate: (cfg, _, currentNode) => {
      currentNode.width = 200;
    },
  });

  test('should get correct width by layoutCoordinate', () => {
    facet.layoutResult.colLeafNodes.forEach((item) => {
      expect(item.width).toBe(200);
    });
  });
});
