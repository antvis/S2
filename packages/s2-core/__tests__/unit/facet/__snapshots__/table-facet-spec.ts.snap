// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Mode Facet Test With Adaptive Layout should get correct col layout col hierarchy coordinate with adaptive layout 1`] = `
Array [
  Object {
    "height": 30,
    "width": 119.8,
    "x": 0,
    "y": 0,
  },
  Object {
    "height": 30,
    "width": 119.8,
    "x": 119.8,
    "y": 0,
  },
  Object {
    "height": 30,
    "width": 119.8,
    "x": 239.6,
    "y": 0,
  },
  Object {
    "height": 30,
    "width": 119.8,
    "x": 359.4,
    "y": 0,
  },
  Object {
    "height": 30,
    "width": 119.8,
    "x": 479.2,
    "y": 0,
  },
]
`;

exports[`Table Mode Facet Test With Adaptive Layout should get correct col layout with seriesNumber col hierarchy coordinate with adaptive layout with seriesNumber 1`] = `
Array [
  Object {
    "height": 30,
    "width": 103.8,
    "x": 80,
    "y": 0,
  },
  Object {
    "height": 30,
    "width": 103.8,
    "x": 183.8,
    "y": 0,
  },
  Object {
    "height": 30,
    "width": 103.8,
    "x": 287.6,
    "y": 0,
  },
  Object {
    "height": 30,
    "width": 103.8,
    "x": 391.40000000000003,
    "y": 0,
  },
  Object {
    "height": 30,
    "width": 103.8,
    "x": 495.20000000000005,
    "y": 0,
  },
]
`;

exports[`Table Mode Facet With Column Grouping Frozen Test should get correct frozenInfo 1`] = `
Object {
  "frozenCol": Object {
    "range": Array [
      0,
      0,
    ],
    "width": 199.66,
    "x": 0,
  },
  "frozenRow": Object {
    "height": 60,
    "range": Array [
      0,
      1,
    ],
    "y": 0,
  },
  "frozenTrailingCol": Object {
    "range": Array [
      2,
      2,
    ],
    "width": 199.66,
    "x": 399.34000000000003,
  },
  "frozenTrailingRow": Object {
    "height": 60,
    "range": Array [
      30,
      31,
    ],
    "y": 502,
  },
}
`;

exports[`Table Mode Facet With Frozen Test should get correct frozenInfo 1`] = `
Object {
  "frozenCol": Object {
    "range": Array [
      0,
      1,
    ],
    "width": 239.6,
    "x": 0,
  },
  "frozenRow": Object {
    "height": 60,
    "range": Array [
      0,
      1,
    ],
    "y": 0,
  },
  "frozenTrailingCol": Object {
    "range": Array [
      3,
      4,
    ],
    "width": 239.60000000000002,
    "x": 359.4,
  },
  "frozenTrailingRow": Object {
    "height": 60,
    "range": Array [
      30,
      31,
    ],
    "y": 502,
  },
}
`;
